packages/cdktf-cli/lib/cdktf-project.ts

```diff
@@ -253,31 +324,89 @@ export class CdktfProject {
     this.stacksToRun.forEach((stack) => stack.stop());
   }
 
+  private waitForApproval() {
+    this.waitingForApproval = true;
+  }
+
+  private resumeAfterApproval() {
+    // We first need to flush all events, we can not resume if there is a new waiting for approval update
+    let event = this.eventBuffer.shift();
+    while (event) {
+      if (event.type === "projectUpdate") {
+        event.cb(event.value);
+        if (event.value.type === "waiting for approval") {
+          // We have to wait for approval again, therefore we can not resume
+          return;
+        }
+      }
+      if (event.type === "logMessage") {
+        event.cb(event.value);
+      }
+
+      event = this.eventBuffer.shift();
+    }
```

Comment:
nit: This could've been a do-while loop, which would get rid of the duplicated `event = this.eventBuffer.shift();` at the bottom.
Despite that, you could also move the `if (event.value.type === "waiting for approval")` into the condition of the do-while loop to stop iterating when you encounter an `waiting for approval` event.
To return without setting waitingForApproval to false, you could check the last event for its type.
Sorry for being so nitty here 😅 I'm unsure whether my proposal is actually easier to understand or not, so I'm happy with you just dismissing this discussion 😁
For comparison, this is what I had in mind:
```ts
// We first need to flush all events, but we can not resume if there is a new waiting for approval update
let event;
do {
  event = this.eventBuffer.shift();
  event.cb(event.value);
} while (event && event.value.type !== "waiting for approval");

if (event.value.type !== "waiting for approval") {
  this.waitingForApproval = false;  
} // else: We have to wait for approval again, therefore we can not resume
```

Reply:
While I don't like what we have I have to say I'd prefer a double  `event = ...` part over checking for the reason we left the loop after the loop was left. I guess this piece of code is what we get for not using RxJS, but on the other hand we don't have to deal with RxJS, so I'd consider it a fair trade :D

Reply:
fine with me, no strong feelings here :)